### PR TITLE
[rfr] Preprints

### PR DESCRIPTION
**Pull request by sloria**
Adds `Preprint`, `Subject`, and `PreprintProvider` models. `Preprint` is a typed model whose field names were changed (similar to Institutions). I will update v1 code to reflect the field changes once we get the views tests running.

The preprint file code is commented out, pending the implementation of StoredFileNode.

Tested by https://github.com/sloria/osf.io/commit/608b0dc4a0c27299d37c460e7b77861c3db334ca

### CODE REVIEW

File: osf_models/models/preprint.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                self.save()
+            return True
+        return False
+
+    def set_preprint_subjects(self, preprint_subjects, auth, save=False):
+        if not self.has_permission(auth.user, ADMIN):
+            raise PermissionsError('Only admins can change a preprint\'s subjects.')
+
+        self.subjects.clear()
+        self.subjects.add(*Subject.objects.filter(guid__object_id__in=preprint_subjects).all())
```
Comment (cwisecarver) [inline osf_models/models/preprint.py]: You should be able to save a little bit of time with:
`self.subjects.add(*Subject.objects.filter(guid__object_id__in=preprint_subjects).values_list('pk', flat=True))`

tl;dr you can pass primary_keys to `add()`

File: osf_models/models/subject.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# -*- coding: utf-8 -*-
+from django.db import models
+
+from website.util import api_v2_url
+
+from osf_models.models.base import BaseModel, ObjectIDMixin
+
+
+class Subject(ObjectIDMixin, BaseModel):
```
Comment (cwisecarver) [inline osf_models/models/subject.py]: This will need the `modm_model_path` and possibly the `modm_query` properties for it to migrate.

File: osf_models/models/preprint_provider.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# -*- coding: utf-8 -*-
+from django.db import models
+
+from website.util import api_v2_url
+
+from osf_models.models.base import BaseModel, ObjectIDMixin
+
+class PreprintProvider(ObjectIDMixin, BaseModel):
```
Comment (cwisecarver) [inline osf_models/models/preprint_provider.py]: This will need the `modm_model_path` and possibly the `modm_query` properties for it to migrate.

**Comment by cwisecarver** [comment]
Other than the `modm_model_path` properties, this lgtm.
